devices/xiaomi.js

```diff
@@ -2564,6 +2595,11 @@ module.exports = [
             exposes.binary('window_open', ea.STATE, true, false),
             e.valve_detection().setAccess('state', ea.ALL),
             e.away_preset_temperature().withAccess(ea.ALL),
+            exposes.switch().withState('schedule', true,
+                'When being ON, the thermostat will change its state based on your settings',
+                ea.ALL, 'ON', 'OFF'),
+            exposes.text('schedule_settings', ea.ALL)
```

Comment:
Can we make this expose a composite or list like: https://github.com/Koenkk/zigbee-herdsman-converters/blob/a588bdeba82fb38bb3d8c7a309779e73744f7609/devices/xiaomi.js#L2592

Reply:
Is there any detailed documentation on that, specifically how to create a list with a fixed number of composite entries, and how to handle the resulting data in `toZigbee`?

Reply:
Figured I can define 4 static composites instead of using the list expose. Is this what you meant?
![image](https://user-images.githubusercontent.com/189372/205268065-cfad5a0f-c227-44de-896d-c4e7008c3055.png)

Is it possible to 
 * change the "days" enum dropdown to a multiselect?
   * alternatively to the dropdown, is it possible to change the "days2" switches to checkboxes and/or position them horizontally so they don't waste so much space?
 * remove all individual "Apply" buttons and have only one for the whole composite? (This is all managed by a single Zigbee message so it doesn't make sense to apply them individually)
 * Add a "get" button to the overall composite instead of the individual properties? (same reason as above)

Still wondering how to submit this form data and process it in the converter.

Reply:
This can get a bit complicated indeed, I'm wondering if the schedule can also set per day/entry? If yes, then we can do something like this: https://github.com/Koenkk/zigbee-herdsman-converters/blob/a069993cd8ca18dd4dc3d9198aa3bb52e7518467/devices/tuya.js#L2077 . In this case there is one expose just for setting the schedule, and below are multiple exposes (read only) showing the schedule per day.

Reply:
Nope, this is all-or-nothing.

Reply:
Can you post the current exposes code?

Reply:
```js
        exposes.composite('schedule_settings_composite')
            .withFeature(exposes.enum('days', exposes.access.STATE_SET, ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'])) // TODO multiselect
            .withFeature(exposes.composite('days2', ea.ALL)
                .withFeature(exposes.switch().withState('mon', true, undefined, ea.STATE_SET))
                .withFeature(exposes.switch().withState('tue', true, undefined, ea.STATE_SET))
                .withFeature(exposes.switch().withState('wed', true, undefined, ea.STATE_SET))
                .withFeature(exposes.switch().withState('thu', true, undefined, ea.STATE_SET))
                .withFeature(exposes.switch().withState('fri', true, undefined, ea.STATE_SET))
                .withFeature(exposes.switch().withState('sat', true, undefined, ea.STATE_SET))
                .withFeature(exposes.switch().withState('sun', true, undefined, ea.STATE_SET)))
            .withFeature(exposes.composite('start', ea.STATE_SET)
                .withFeature(exposes.numeric('hour', ea.STATE_SET))
                .withFeature(exposes.numeric('minute', ea.STATE_SET))
                .withFeature(exposes.numeric('temperature', ea.STATE_SET)))
            .withFeature(exposes.composite('intermediate1', ea.STATE_SET)
                .withFeature(exposes.numeric('hour', ea.STATE_SET))
                .withFeature(exposes.numeric('minute', ea.STATE_SET))
                .withFeature(exposes.numeric('temperature', ea.STATE_SET)))
            .withFeature(exposes.composite('intermediate1', ea.STATE_SET)
                .withFeature(exposes.numeric('hour', ea.STATE_SET))
                .withFeature(exposes.numeric('minute', ea.STATE_SET))
                .withFeature(exposes.numeric('temperature', ea.STATE_SET)))
            .withFeature(exposes.composite('end', ea.STATE_SET)
                .withFeature(exposes.numeric('hour', ea.STATE_SET))
                .withFeature(exposes.numeric('minute', ea.STATE_SET))
                .withFeature(exposes.numeric('temperature', ea.STATE_SET)))
            .withDescription('Smart schedule configuration'),
```

Reply:
Thanks, I'm wondering wether this should become a list (this seems more logical to me). Similar to https://github.com/Koenkk/zigbee-herdsman-converters/blob/a588bdeba82fb38bb3d8c7a309779e73744f7609/devices/xiaomi.js#L2592 

So the list contains the schedule entries, each schedule entry contains a day, time and temperature.

Reply:
See my question above:

> Is there any detailed documentation on [...] how to create a list **with a fixed number of composite entries**?

This TRV has a very basic scheduler implementation that requires a fixed amount of 4 time/temperature entries, and the current list implementation doesn't seem to allow to require a fixed number - or did I miss this? Also, the schedule only allows a global selection of days - not multiple configurations for multiple days. Please check the description of this PR, it documents exactly how the schedule settings look like.

Reply:
> Also, the schedule only allows a global selection of days - not multiple configurations for multiple days

Can't this be achieved by wrapping the `days` in a list?

@nurikk we need your help here.

> requires a fixed amount of 4 time/temperature entries

For the list type we can add a `min`/`max` entries, then for this usecase we set both to 4. Could this be supported in the frontend?

Reply:
> Can't this be achieved by wrapping the `days` in a list?

Do you mean like this?

```js
exposes.composite('scheduleTest')
  .withFeature(exposes.list('days', ea.STATE_SET, 
    exposes.enum('days', exposes.access.STATE_SET, ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'])))
```
This would allow invalid or at undesired combinations like `['wed', 'tue', 'mon']` (unexpected order) or `['tue', 'tue']` (unexpected duplicates). I know it can be filtered, but it still creates an odd user experience. The day selection needs something like this
![image](https://user-images.githubusercontent.com/189372/205590101-a5baa1aa-ecf0-4acc-bd9c-4a268d9ef8ba.png)
or this
![image](https://user-images.githubusercontent.com/189372/205590148-3e2bfc49-01b2-435b-aa9b-e0bd92c88dbd.png)

Reply:
That would be better indeed, @nurikk can this be implemented in the frontend?

Reply:
@nurikk excuse me, what did you say? What exactly should be supported, and where?

Reply:
What is a "link expose"? The state you are looking for is documented in the PR description.

Reply:
sorry, my mobile spelling sucks. I meant `list exposes feature`

Reply:
Min/max, can be supported. Kindly raise an issue in zigbee2mqtt-frontend issue tracker, I’ll check it this weekend

Reply:
So does that mean that the mockups cannot be realized?

Reply:
@protyposis can you provide me an example of what you publish over MQTT to set `schedule_settings_json`?

Reply:
An example is in the PR description:

> ```json
>  {
>      "repeat": ["mon", "tue", "wed", "thu", "fri", "sat", "sun"],
>      "events": [
>          {"time": 480, "temperature": 24},
>          {"time": 1080, "temperature": 17.3},
>          {"time": 1380, "temperature": 22},
>          {"time": 480, "temperature": 22},
>      ],
>  }
> ```

To accommodate it to the proposed composite structure in https://github.com/Koenkk/zigbee-herdsman-converters/pull/5058/files#discussion_r1038766761, I tried it like this:

```js
{
  days: { mon: 'ON', tue: 'ON', wed: 'ON', thu: 'ON', fri: 'ON', sat: 'ON', sun: 'ON' },
  start: { hour: 8, minute: 0, temperature: 24 },
  intermediate1: { hour: 18, minute: 0, temperature: 17.3 },
  intermediate2: { hour: 23, minute: 0, temperature: 22 },
  end: { hour: 8, minute: 0, temperature: 22 },
}
```

Reply:
If I understood correctly, this is the full and only schedule, right?

e.g. is it possible to set a different schedule for a day? e.g.

```
 {
     "repeat": ["mon", "tue", "wed", "thu", "fri"],
     "events": [
         {"time": 480, "temperature": 24},
         {"time": 1080, "temperature": 17.3},
         {"time": 1380, "temperature": 22},
         {"time": 480, "temperature": 22}
     ],
 }
```

```
 {
     "repeat": ["sat", "sun"],
     "events": [
         {"time": 520, "temperature": 24},
         {"time": 1500, "temperature": 22}
     ],
 }
```

Reply:
Thanks, I will check this with @nurikk

Reply:
@protyposis wanted to let you know I did not forget about this, working in https://github.com/nurikk/zigbee2mqtt-frontend/pull/1638

Reply:
Great, thanks @Koenkk. Will follow up here when https://github.com/nurikk/zigbee2mqtt-frontend/pull/1638 is finalized.

Reply:
Apologies for taking so long, but all should be good now (in the latest dev). 

You can use the following expose:

```js
exposes.composite('Schedule', 'schedule', ea.ALL)
    .withFeature(exposes.composite('Days', 'days')
        .withFeature(exposes.switch().withState('mon', true, undefined, ea.STATE))
        .withFeature(exposes.switch().withState('tue', true, undefined, ea.STATE))
        .withFeature(exposes.switch().withState('wed', true, undefined, ea.STATE))
        .withFeature(exposes.switch().withState('thu', true, undefined, ea.STATE))
        .withFeature(exposes.switch().withState('fri', true, undefined, ea.STATE))
        .withFeature(exposes.switch().withState('sat', true, undefined, ea.STATE))
        .withFeature(exposes.switch().withState('sun', true, undefined, ea.STATE)),
    )
    .withFeature(exposes.composite('Start', 'start')
        .withFeature(exposes.numeric('Hour', ea.STATE).withProperty('hour'))
        .withFeature(exposes.numeric('Minute', ea.STATE).withProperty('minute'))
        .withFeature(exposes.numeric('Temperature', ea.STATE).withProperty('temperature')))
    .withFeature(exposes.composite('Intermediate 1', 'intermediate1')
        .withFeature(exposes.numeric('Hour', ea.STATE).withProperty('hour'))
        .withFeature(exposes.numeric('Minute', ea.STATE).withProperty('minute'))
        .withFeature(exposes.numeric('Temperature', ea.STATE).withProperty('temperature')))
    .withFeature(exposes.composite('Intermediate 2', 'intermediate2')
        .withFeature(exposes.numeric('Hour', ea.STATE).withProperty('hour'))
        .withFeature(exposes.numeric('Minute', ea.STATE).withProperty('minute'))
        .withFeature(exposes.numeric('Temperature', ea.STATE).withProperty('temperature')))
    .withFeature(exposes.composite('End', 'end')
        .withFeature(exposes.numeric('Hour', ea.STATE).withProperty('hour'))
        .withFeature(exposes.numeric('Minute', ea.STATE).withProperty('minute'))
        .withFeature(exposes.numeric('Temperature', ea.STATE).withProperty('temperature')))
    .withDescription('Smart schedule configuration'),
```

On you use this and submit it through the z2m frontend -> exposes tab, you will see the structure in the z2m debug log.

Reply:
On the latest HA dev version I get `N/A` everywhere. if I change the `ea.STATE` to `ea.ALL`, I get the input fields but with reload buttons everywhere again. Any idea what's wrong?
![image](https://user-images.githubusercontent.com/189372/212736954-f9f9037e-0b26-42e5-b500-0f4086a40444.png)

Reply:
Is the property in the correct structure? (can you provide the debug log when it starts showing `N/A`)

See https://www.zigbee2mqtt.io/guide/usage/debug.html on how to enable debug logging.

Reply:
Right, the `N/A` issue is fixed by supplying the correct structure, still there are 3 issues left:
 1. I need to use `ea.STATE_SET` instead of your suggested `ea.STATE` to get editable input fields instead of read-only text labels. Is this ok?
 2. When editing fields, other fields get unexpectedly cleared. For example, if I click into the `minute` field in `start` and change the value, the `hour` and `temperature` get unexpectedly cleared (emptied). Same behaviour in the other 3 time/temperature composites.
 3. The `Apply` button never gets enabled, even when all fields are filled, so the schedule can never be submitted.

To be perfectly honest, this PR is getting out of hand and I would be very much in favor if I could just clean it up and we merge it with the text configuration I initially proposed (it wouldn't be the only device with a text-config). If there's demand, which we don't even know yet, someone can later handle the UI aspect in a separate PR and make it more user friendly. Could you live with this iterative approach?

Reply:
1. Yes that is fine
2. I cannot reproduce it, any chance you create a GIF for this?
3. Same as 2, can you also make a GIF of this? It should be enabled when clicking next to the text field after filling in the last value.


> To be perfectly honest, this PR is getting out of hand

I agree, I'm also fine in accepting it with just a text expose but with the JSON structure that this expose will send. Because when this gets merged, we cannot change the format anymore (since it will give a breaking change).

Reply:
Here's a screen recording in Chrome containing both issues (disabled `Apply` button at end of video). Just changing focus from one text field to another already clears the values.

![Recording 2023-01-24 at 08 10 00](https://user-images.githubusercontent.com/189372/214232819-bfd7b9ca-0fa4-4ee7-8002-ee8f3a20fd2f.gif)

> when this gets merged, we cannot change the format anymore (since it will give a breaking change)

Can't the value be upgraded later with a type check, e.g. 
```js
if (typeof value === 'string') {
  // Upgrade old string to new JSON representation to avoid breaking change
  value = upgradeStringToJson(value);
}```
?

Reply:
Could it be that the thermostats publishes the value while changing this? (then this should solved by changing `exposes.composite('schedule'` to `exposes.composite('schedule_set'`, then change `ea.STATE_SET` to `ea.SET`

> Can't the value be upgraded later in convert or configure with a type check, e.g.

I'm fine with that.

Reply:
This unfortunately doesn't fix it either, I'm giving up on the UI for now and will revert to the configuration string.

Reply:
Agree!